src/neural/wrapper.cc

```diff
@@ -0,0 +1,148 @@
+    attrs_.suggested_num_search_threads = network_->GetThreads();
+    attrs_.recommended_batch_size = network_->GetMiniBatchSize();
+    attrs_.maximum_batch_size = 1024;
+    input_format_ = caps.input_format;
+  }
+
+  BackendAttributes GetAttributes() const override { return attrs_; }
+  virtual std::unique_ptr<BackendComputation> CreateComputation() override;
+
+ private:
+  std::unique_ptr<Network> network_;
+  BackendAttributes attrs_;
+  pblczero::NetworkFormat::InputFormat input_format_;
+  float softmax_policy_temperature_;
+  friend class NetworkAsBackendComputation;
+};
+
+class NetworkAsBackendComputation : public BackendComputation {
+ public:
+  NetworkAsBackendComputation(NetworkAsBackend* backend)
+      : backend_(backend), computation_(backend_->network_->NewComputation()) {
+    results_.reserve(backend_->attrs_.maximum_batch_size);
+    moves_.reserve(backend_->attrs_.maximum_batch_size);
+    transforms_.reserve(backend_->attrs_.maximum_batch_size);
+  }
+
+  size_t UsedBatchSize() const override { return computation_->GetBatchSize(); }
+
+  AddInputResult AddInput(const EvalPosition& pos,
+                          EvalResultPtr result) override {
+    int transform;
+    computation_->AddInput(EncodePositionForNN(backend_->input_format_, pos.pos,
+                                               8, FillEmptyHistory::FEN_ONLY,
+                                               &transform));
+    results_.push_back(result);
+    moves_.emplace_back(pos.legal_moves.begin(), pos.legal_moves.end());
+    transforms_.push_back(transform);
+    return ENQUEUED_FOR_EVAL;
+  }
+
+  void ComputeBlocking() override {
+    computation_->ComputeBlocking();
+    for (size_t i = 0; i < results_.size(); ++i) {
+      const EvalResultPtr& result = results_[i];
+      if (result.q) *result.q = computation_->GetQVal(i);
+      if (result.d) *result.d = computation_->GetDVal(i);
+      if (result.m) *result.m = computation_->GetMVal(i);
+      if (!result.p.empty()) SoftmaxPolicy(result.p, computation_.get(), i);
+    }
+  }
+
+  void SoftmaxPolicy(std::span<float> dst,
+                     const NetworkComputation* computation, int idx) {
+    const std::vector<Move>& moves = moves_[idx];
+    const int transform = transforms_[idx];
+    // Copy the values to the destination array and compute the maximum.
+    const float max_p = std::accumulate(
+        moves.begin(), moves.end(), -std::numeric_limits<float>::infinity(),
+        [&, counter = 0](float max_p, const Move& move) mutable {
+          return std::max(max_p, dst[counter++] = computation->GetPVal(
+                                     idx, move.as_nn_index(transform)));
+        });
+    // Compute the softmax and compute the total.
+    const float temperature = backend_->softmax_policy_temperature_;
+    float total = std::accumulate(
+        dst.begin(), dst.end(), 0.0f, [&](float total, float& val) {
+          return total + (val = FastExp((val - max_p) * temperature));
```

Comment:
Not a review comment but a note to myself to check whether `FastExp()` helps here.

Reply:
I decided to keep it for now and then change it in a separate step when everything is more stable, so that if the performance degrades after the switch (which is likely), it would be easier to find a cause.